### DISCUSSION TITLE
[Implementation-Review] Tacotron 1 - CBHG module - Max pooling missing (?)

### DISCUSSION THREAD

**Discussion by p0p4k** [coqui-ai/TTS#2115] [General Q&A] [answered]
Hello,
If you notice another popular open source implementation of [Tacotron](https://github.com/keithito/tacotron/blob/d26c763342518d4e432e9c4036a1aff3b4fdaa1e/models/modules.py#L47-L52), they implement the max-pooling as advised in the paper ([pdf link](https://arxiv.org/pdf/1703.10135.pdf)). 
```
    # Maxpooling:
    maxpool_output = tf.layers.max_pooling1d(
      conv_outputs,
      pool_size=2,
      strides=1,
      padding='same')
```
Whereas in our side [implementation](https://github.com/coqui-ai/TTS/blob/fa0e71d0b694d15f3e1885667ec5cfbfa827f549/TTS/tts/layers/tacotron/tacotron.py#L145-L146), right after `conv1d_banks`, we use the `conv1d_projections`. I know it does not make any major difference, but is there anything I am missing? Thanks.

```
    def forward(self, inputs):
        # (B, in_features, T_in)
        x = inputs
        # (B, hid_features*K, T_in)
        # Concat conv1d bank outputs
        outs = []
        for conv1d in self.conv1d_banks:
            out = conv1d(x)
            outs.append(out)
        x = torch.cat(outs, dim=1)
        assert x.size(1) == self.conv_bank_features * len(self.conv1d_banks)
        for conv1d in self.conv1d_projections:
            x = conv1d(x)
        x += inputs
        x = x.transpose(1, 2)
        if self.highway_features != self.conv_projections[-1]:
            x = self.pre_highway(x)
        # Residual connection
        # TODO: try residual scaling as in Deep Voice 3
        # TODO: try plain residual layers
        for highway in self.highways:
            x = highway(x)
        # (B, T_in, hid_features*2)
        # TODO: replace GRU with convolution as in Deep Voice 3
        self.gru.flatten_parameters()
        outputs, _ = self.gru(x)
        return outputs
```

**Comment by erogol** [comment, marked as answer]
I don't remember why we switch to projection layers but I remember cautiously we switched to projection layers at some point. But yeah that part is different than the paper.

**Reply by p0p4k** [reply]
Even in the paper they do projections, but there is pooling between them. It does not make a big difference at the end of the day , as backprop will find a solution anyway.

**Reply by erogol** [reply]
Hope so :)
